In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [2]:
X_train = pd.read_csv('https://raw.githubusercontent.com/simsekergun/ENEE691/main/homework3/Reflectance_X_train.csv')
Y_train = pd.read_csv('https://raw.githubusercontent.com/simsekergun/ENEE691/main/homework3/Reflectance_Y_train.csv')
X_test = pd.read_csv('https://raw.githubusercontent.com/simsekergun/ENEE691/main/homework3/Reflectance_X_test.csv')
n_train = Y_train.iloc[:,1]
k_train = Y_train.iloc[:,2]

X_train.shape, Y_train.shape, X_test.shape

((3060, 244), (3060, 3), (1, 244))

In [3]:
nk_train = Y_train.iloc[:,[1,2]]

### TASK 1

In [4]:
from sklearn.linear_model import LinearRegression  

reg = LinearRegression()         # Instantiate
reg.fit(X_train, n_train)        # Fit the train data

r2_train_score = reg.score(X_train, n_train)   # Calculating R^2 score for train

print('R^2 score for n-train dataset = ', round(r2_train_score, 4), '\n')

R^2 score for n-train dataset =  0.9965 



In [5]:
n_pred = reg.predict(X_test)
n_pred

array([2.51788926])

In [6]:
regk = LinearRegression()         # Instantiate
regk.fit(X_train, k_train)        # Fit the train data

r2k_train_score = regk.score(X_train, k_train)   # Calculating R^2 score for train

print('R^2 score for k-train dataset = ', round(r2k_train_score, 4), '\n')

k_pred = regk.predict(X_test)
k_pred

R^2 score for k-train dataset =  0.9948 



array([3.68006685])

### TASK 2

In [7]:
from sklearn.linear_model import LogisticRegression 

In [8]:
# let's convert n and k values to integers after multiplying them 10
ns = (n_train*10).astype(int)
ks = (k_train*10).astype(int)
ns

0        1
1        1
2        1
3        1
4        1
        ..
3055    60
3056    60
3057    60
3058    60
3059    60
Name: n, Length: 3060, dtype: int64

In [9]:
logreg = LogisticRegression(max_iter = 100000) # Instantiating logistic regression 
logreg.fit(X_train, ns)  # Fitting the train data to 'logreg'
y_test_pred_prob = logreg.predict_proba(X_test) 

In [10]:
y_test_pred_prob

array([[8.73665366e-26, 2.11394163e-22, 1.63785099e-18, 2.43711063e-15,
        6.19439660e-13, 6.27644424e-11, 3.16907247e-09, 8.35668783e-08,
        1.21222304e-06, 1.07857727e-05, 6.31141286e-05, 2.60331081e-04,
        8.14873572e-04, 2.07094166e-03, 4.43895600e-03, 8.23104776e-03,
        1.37330879e-02, 2.08766576e-02, 2.91754739e-02, 3.82513321e-02,
        4.74507377e-02, 5.58410938e-02, 6.28898900e-02, 6.81412519e-02,
        7.11342602e-02, 7.18263381e-02, 7.03230423e-02, 6.69444981e-02,
        6.20282529e-02, 5.60440843e-02, 4.95027732e-02, 4.26775522e-02,
        3.58855217e-02, 2.95047984e-02, 2.37819064e-02, 1.87513266e-02,
        1.44185105e-02, 1.08276808e-02, 7.90539009e-03, 5.60430895e-03,
        3.87367201e-03, 2.59161611e-03, 1.68271221e-03, 1.05667929e-03,
        6.36201237e-04, 3.65314889e-04, 1.98751203e-04, 1.01552662e-04,
        4.83683461e-05, 2.12510026e-05, 8.49175524e-06, 3.02962050e-06,
        9.39348152e-07, 2.43919858e-07, 4.98902284e-08, 7.395535

In [11]:
nvalues = np.linspace(0.1,6,60)
df = pd.DataFrame(y_test_pred_prob, columns = nvalues).T
nfinalists = df.nlargest(4,0)
nfinalists

0
2.6  0.071826
2.5  0.071134
2.7  0.070323
2.4  0.068141

In [12]:
nf = (2.6*0.071826+2.5*0.071134+2.7*0.070323+2.4*0.068141)/(0.071826+0.071134+0.070323+0.068141)
nf

2.5512859599749844

In [13]:
logregk = LogisticRegression(max_iter = 100000) # Instantiating logistic regression 
logregk.fit(X_train, ks)  # Fitting the train data to 'logreg'

LogisticRegression(max_iter=100000)

In [14]:
y_test_pred_prob2 = logregk.predict_proba(X_test) 

In [15]:
kvalues = np.linspace(0.0,5,51)
dfk = pd.DataFrame(y_test_pred_prob2, columns = kvalues).T
kfinalists = dfk.nlargest(4,0)
kfinalists

0
3.6  0.091211
3.7  0.089014
3.5  0.088773
3.4  0.082375

In [16]:
kf = (3.6*0.091211+3.7*0.089014+3.5*0.088773+3.4*0.082375)/(0.091211+0.089014+0.088773+0.082375)
kf

3.553181092457303

Linear regression prediction: 2.51789+3.68i <br>
Regression via LogReg prediction: 2.55129 + 3.553i  <br>
They are close enough to each other. We can claim that we achieved regression via logistic regression.

### TASK 3

In [18]:
from sklearn import svm
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report

In [22]:
def svm_clf(X_tr, y_tr, ker, C, gamma='auto', r=0, deg=1):
    clf = svm.SVC(kernel=ker, gamma=gamma, C=C, coef0=r, degree=deg)
    clf.fit(X_tr, y_tr)              # Fit the train data
    y_tr_pred = clf.predict(X_tr)    # Predict the target for train data
    tr_score = accuracy_score(y_tr, y_tr_pred)  # Compute train accuracy    
    print('Train R Score: \n', tr_score,'\n')
    return 

In [23]:
# Apply linear SVM with very low budget for margin violation
svm_clf(X_train, ns, ker='linear', C=0.01)

Train R Score: 
 0.07254901960784314 



In [24]:
# Apply linear SVM with very high budget for margin violation
svm_clf(X_train, ns, ker='linear', C=100)

Train R Score: 
 0.9630718954248366 



R2 of 96% is quite satisfactory. Let's make a prediction.

In [30]:
clf = svm.SVC(kernel='linear', gamma='auto', C=100, probability=True)
clf.fit(X_train, ns)              # Fit the train data


SVC(C=100, gamma='auto', kernel='linear', probability=True)

In [32]:
probabilities = clf.predict_proba(X_test)
probabilities

array([[8.66259914e-04, 3.59450057e-04, 3.37068999e-04, 2.95396420e-04,
        3.12172027e-04, 5.46688307e-04, 5.00614222e-04, 4.74033810e-04,
        5.48736257e-04, 8.32596017e-04, 1.23298618e-03, 1.82491438e-03,
        3.58787422e-03, 4.44420003e-03, 3.83179690e-03, 2.87262536e-03,
        3.38510111e-03, 4.28498937e-03, 5.51491797e-03, 7.47640348e-03,
        9.98226943e-03, 1.25491054e-02, 1.58792993e-02, 3.29553073e-02,
        8.31361675e-01, 1.36615109e-02, 5.04795418e-03, 3.33206887e-03,
        2.47556749e-03, 1.99614216e-03, 1.68486219e-03, 1.48667615e-03,
        1.35323284e-03, 1.25789602e-03, 1.18426297e-03, 1.12011775e-03,
        1.07550176e-03, 1.04559612e-03, 1.02725499e-03, 9.96985045e-04,
        9.59115999e-04, 9.26565063e-04, 9.00905879e-04, 8.74485982e-04,
        8.48643499e-04, 8.25049703e-04, 8.04846882e-04, 7.83067683e-04,
        7.63902782e-04, 7.46112202e-04, 7.26799469e-04, 7.10568703e-04,
        6.93814200e-04, 6.77834637e-04, 6.61546388e-04, 6.468711

In [33]:
nvalues = np.linspace(0.1,6,60)
df = pd.DataFrame(probabilities, columns = nvalues).T
nfinalists = df.nlargest(4,0)
nfinalists

0
2.5  0.831362
2.4  0.032955
2.3  0.015879
2.6  0.013662

In [35]:
nf = (2.6*0.013662+2.5*0.831362+2.3*0.015879+2.4*0.032955)/(0.013662+0.831362+0.015879+0.032955)
nf

2.49428869014989

In [36]:
clf2 = svm.SVC(kernel='linear', gamma='auto', C=100, probability=True)
clf2.fit(X_train, ks)              # Fit the train data

SVC(C=100, gamma='auto', kernel='linear', probability=True)

In [37]:
probabilities2 = clf2.predict_proba(X_test)
kvalues = np.linspace(0.0,5,51)
dfk = pd.DataFrame(probabilities2, columns = kvalues).T
kfinalists = dfk.nlargest(4,0)
kfinalists

0
3.7  0.464755
3.8  0.292032
3.6  0.033087
3.9  0.030839

In [38]:
kf = (3.6*0.033087+3.7*0.464755+3.8*0.292032+3.9*0.030839)/(0.033087+0.464755+0.292032+0.030839)
kf

3.739066397144922

Linear regression prediction: 2.51789+3.68i <br>
Regression via LogReg prediction: 2.55129 + 3.553i <br>
Regression via SVM prediction:  2.494+3.739i<br>

They are all very close to each other. We can claim that we achieved regression via SVM as well.

SVM took longer than the logistic regression.


The true (experimentally measured) refractive index is 2.5 + 3.74i. <br>

Linear Regression Error (real part): 0.70% <br>
Linear Regression Error (imag. part): 1.6% <br>
<br>
LogReg Regression Error (real part): 2% <br>
LogReg Regression Error (imag. part): 5% <br>
<br>
SVM Regression Error (real part): 0.23% <br>
SVM Regression Error (imag. part): 0.02% <br>
<br>
As a result, we can claim that SVM based regression is the ultimate winner despite its slowness